service/src/main/java/bio/terra/axonserver/service/file/FileService.java

```diff
@@ -0,0 +1,93 @@
+import bio.terra.axonserver.service.exception.InvalidResourceTypeException;
+import bio.terra.axonserver.service.iam.AuthenticatedUserRequest;
+import bio.terra.axonserver.service.wsm.WorkspaceManagerService;
+import bio.terra.axonserver.utils.GcpUtils;
+import bio.terra.workspace.model.ResourceDescription;
+import com.google.auth.oauth2.GoogleCredentials;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import org.apache.commons.io.FilenameUtils;
+import org.javatuples.Pair;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.ByteArrayResource;
+import org.springframework.stereotype.Component;
+
+@Component
+public class FileService {
+
+  WorkspaceManagerService wsmService;
+  ConvertService convertService;
+
+  @Autowired
+  public FileService(WorkspaceManagerService wsmService, ConvertService convertService) {
+    this.wsmService = wsmService;
+    this.convertService = convertService;
+  }
+
+  public ByteArrayResource getFile(
+      AuthenticatedUserRequest userRequest,
+      UUID workspaceId,
+      UUID resourceId,
+      @Nullable String objectPath,
+      @Nullable String convertTo) {
+
+    ResourceDescription resource = this.getResource(userRequest, workspaceId, resourceId);
+
+    Pair<byte[], String> fileWithName = this.getFileHandler(resource, objectPath, userRequest);
+    byte[] file = fileWithName.getValue0();
+    String fileName = fileWithName.getValue1();
+
+    if (convertTo != null) {
+      String fileExtension = FilenameUtils.getExtension(fileName);
+      file = this.convertService.convert(file, fileExtension, convertTo, userRequest);
+    }
+    return new ByteArrayResource(file);
+  }
+
+  private Pair<byte[], String> getFileHandler(
+      ResourceDescription resource,
+      @Nullable String objectPath,
+      AuthenticatedUserRequest userRequest) {
+
+    return switch (resource.getMetadata().getResourceType()) {
+      case GCS_OBJECT -> {
+        String objectName = resource.getResourceAttributes().getGcpGcsObject().getFileName();
+        yield new Pair<byte[], String>(this.getGcsObjectFile(resource, userRequest), objectName);
+      }
+      case GCS_BUCKET -> new Pair<byte[], String>(
+          this.getGcsBucketFile(resource, objectPath, userRequest), objectPath);
+      default -> throw new InvalidResourceTypeException(
+          "Not a file containing resource: " + resource.getMetadata().getResourceType());
+    };
+  }
+
+  private byte[] getGcsObjectFile(
+      ResourceDescription resource, AuthenticatedUserRequest userRequest) {
+    GoogleCredentials googleCredentials = GcpUtils.getGoogleCredentialsFromUserRequest(userRequest);
+
+    String bucketName = resource.getResourceAttributes().getGcpGcsObject().getBucketName();
+    String objectPath = resource.getResourceAttributes().getGcpGcsObject().getFileName();
+    return new CloudStorageService()
+        .getBucketObject(googleCredentials, bucketName, objectPath)
+        .getContent();
```

Comment:
Was thinking about this and wasn't sure what the best practice is. Currently the UI doesn't do this check, but that's probably fine because its client side. I had looked into using files but I think thats unsafe because temp files being created and also the endpoint needs to be RESTful.

Perhaps a good approach is reading into a buffer of a set size and throwing a "too large" error if the fize is too big?